The 1st draft of score repository structure proposal (#1475)
eclipse-score/score: docs/features/infrastructure/ci_repo_structure_draft.md

> ## 📘 Project Repository Structure Proposal  

Reviewer: this should not be named "Repository Structure Proposal" IMHO

### *Introduction & Motivation*

To address the challenges of S-CORE project — such as slower development cycles, inefficient testing, and governance complexity — we propose a multirepo architecture built on Bazel build automation tool with central configuration repository. In this design, each functional unit resides in its own repository, enabling teams to work independently and iterate quickly.

> A central configuration repository serves as the coordinator. Rather than containing application logic, it defines dependency relationships between modules and acts as the single source for system-wide integration. This setup ensures that only well-tested and validated module versions enter the production build context.

Author: > Shouldn't this be an rst file?
- No, this was not supposed to be presented in this form but TL requested that we publish it anyway. Proper document still needs to be written. 
> So far we did not consider the integration into the a specific project. So there will be bazel modules, which need to be provided be the environment and need to be exchangeable. Don't know if we have a solution for this already e.g. via labelflags.
- I'm sorry, I don't really understand what is the question here.

> A central configuration repository serves as the coordinator. Rather than containing application logic, it defines dependency relationships between modules and acts as the single source for system-wide integration. This setup ensures that only well-tested and validated module versions enter the production build context.

Reviewer 2: let's suppose that we do the reference integration on qemu against a sdp8 from qnx everywhere as described [here](https://github.com/eclipse-score/toolchains_qnx). But in the series project we have a commercial license of qnx. The question is how do we support this use case where  have to replace the bazel targets of the qnx everywhere with the ones from the commercial version. 
But as said above we don't have to answer this right now.

> A central configuration repository serves as the coordinator. Rather than containing application logic, it defines dependency relationships between modules and acts as the single source for system-wide integration. This setup ensures that only well-tested and validated module versions enter the production build context.

Reviewer 3: I would say you need an internal ci/cd anyway with different targets

The combination of modular development and centralized governance creates a scalable and maintainable system. Root-level CI with GitHub Actions further guarantees integration quality without slowing down individual module workflows.

- 🔲 **Scalability Through Modularity**  

>   As project grow, tightly coupled monoliths slow down iteration. By segmenting the codebase into independent Bazel modules, each team can develop, test, and deploy in isolation—improving scalability and parallelism across the organization.

Reviewer 4: ```suggestion
  As projects grow, tightly coupled monoliths slow down iteration. By segmenting the codebase into independent Bazel modules, each team can develop, test, and deploy in isolation - improving scalability and parallelism across the organization.
```

  The central configuration repository contains no application code but acts as the orchestration layer. It defines the dependency graph using Bazel and ensures that modules interact through well-defined, versioned interfaces.

- ⚗️ **Efficient CI Pipeline Design**  
  CI workflows are simplified by running full system integration tests only in the root repository using GitHub Actions. This avoids redundant CI overhead in module repos while ensuring comprehensive validation in the final build.

- 🛡️ **Consistency in Governance**  
  A gated integration process is enforced. Module repositories must pass all defined quality checks before their changes are accepted in the root repository (bump in S-CORE Bazel Registry must be verified by root repo). This maintains system integrity without stifling local agility.

---

## 🏗️ High-Level Multirepo Architecture  
### *Designing for Modularity and Control*

The architecture uses one central repo for Bazel coordination and multiple surrounding module repos. Each module is a Bazel unit. The root repo runs integration builds/tests, ensuring system consistency without blocking individual module progress.

> - ⚙️ **Central Config Repo**  

Reviewer 4: Why not
```suggestion
- ⚙️ **Central Release Repo**  
```
?
Since it's only purpose is to collect modules in their respective version for each release, did I get this right?

  > NOTE: It's possible to have several modules under one git repository to reduce the system complexity, but this has to be approved by Project Structure Expert Group.

- 🧭 **Root Repo Integration Layer**  
  The root repo imports external Bazel modules and tests them together to verify the integrity of the entire system. Integration logic and validation reside here.

- 🤖 **GitHub Actions for System Testing**  
  All full builds and integration tests run via GitHub Actions in the root repo, ensuring module compatibility without duplicating CI across module repos.

---

## 📁 Root Repository Structure  
### *Centralized Bazel Management*

- 🗂️ **MODULE.bazel Management**  
  The root repo defines Bazel modules using `bazel_dep()`, locking specific versions and handling updates in a controlled manner. This prevents accidental mismatches across modules.

- 🔐 **Version Locking & Isolation**  
  The configuration ensures that only compatible versions of modules are referenced together, improving reproducibility and avoiding dependency hell.

- 🔧 **Global Toolchains and Rules**  

>   Bazel toolchains (e.g., Go, Python, C++) and common rules are configured in independent modules, but with pinpointed versions to ensuring consistency in how builds and tests are executed.

Reviewer 4: ```suggestion
  Bazel toolchains (e.g., Go, Python, C++) and common rules are configured in independent modules, but with pinpointed versions for ensuring consistency in how builds and tests are executed.
```

- 💡 **Fail-Safe Guards**  
  Preventative checks are added to ensure all builds are executed from the root and not from within a module (need to see the mechanics for it).

---

## 📦 Module Repository Design  
### *Autonomous, Focused Development*

- 🧰 **Self-Contained Build Targets**  
  Each module defines its own Bazel `BUILD` and `MODULE.bazel` files. Developers can work independently without needing full system awareness.

- 🧪 **Independent CI Pipelines**  
  Modules run their own linting, testing, and coverage steps in isolation, ensuring correctness before integration into the system.

- 🧭 **Cross-Repo Dependencies**  

>   Modules can depend on other modules however root repository will override all versioning down the tree. It's up to module owners to ensure the compatibility in overall system.

Reviewer 4: ```suggestion
  In general, modules can depend on other modules. However, the root repository will override all versioning down the tree. It's up to module owners to ensure the compatibility within the overall system.
```

- 🧹 **Minimal Overhead**  
  Without unnecessary shared tooling or entanglement, the module repos stay lightweight and easier to maintain.

---

## 🧩 Dependency Management with Bzlmod  
### *Composability and Reusability*

- 🔄 **bazel_dep() for Version Control**  
  Modules declare dependencies in `MODULE.bazel` using `bazel_dep()`, enabling clear and repeatable versioning.

- 🪢 **MODULE.bazel Inheritance**  
  Bazel’s Bzlmod system allows modules to compose cleanly via recursive resolution of `MODULE.bazel` dependencies.

- 🛠️ **Extensions and Overrides**  
  Advanced features like toolchain extensions or module overrides enable fine-tuned control of the build graph without modifying upstream repos.

- 🔁 **Update via Tag or SHA**  

>   Module versions are pinned to specific Git tags or release packages to ensure stability and auditability.

Author: I'm not sure if the highlighted reference here is correct. Bazel version will be set by the root module (reference integration). Each module can use version that they prefer as long as there are no combability issues between root version and module version. In the end, root module will override the Bazel version set in module .bazelversion file.

---

## 📦 Dependency Vendoring & Reproducibility  
### *Ensuring Reliable and Offline Builds*

- 🗃️ **Vendored External Dependencies**  
  Bazel supports vendoring external dependencies using `bazel mod vendor`. This exports all transitive module content into a local `vendor/` directory, enabling developers to build the full project without accessing remote repositories.

- 🔐 **Reproducibility with Lock Files**  
  The `MODULE.bazel.lock` file captures a frozen dependency graph. Combined with vendoring, it guarantees builds are consistent across environments and CI systems, regardless of upstream changes.

- 🚫 **Offline and Air-Gapped Compatibility**  
  Teams working in isolated or secure environments can still perform full builds by relying solely on the vendored content, bypassing the need for network access.

- 🧪 **Local Development with `local_path_override()`**  

>   Developers testing changes to a module locally—before publishing it to the Bazel Registry—can override its reference using the `local_path_override()` directive in the root repo’s `MODULE.bazel`. This enables rapid integration testing and reduces friction in multi-repo workflows.

Reviewer 4: ```suggestion
  Developers testing changes to a module locally (before publishing it to the Bazel Registry) can override its reference using the `local_path_override()` directive in the root repo’s `MODULE.bazel`. This enables rapid integration testing and reduces friction in multi-repo workflows.
```

---

## ⚙️ CI/CD via GitHub Actions  
### *Integration Validation at the Root*

- 🧪 **Root Runs Full Integration Tests**  
  When module versions are updated in the root, GitHub Actions runs `bazel build //...` and `bazel test //...` to validate the full system.

- 🚦 **PR Gatekeepers**  

>   Root repo pull requests require CI to pass before merging, serving as a gate to prevent regressions. To save resources only reviewed PRs will go through integration testing.

Reviewer 4: ```suggestion
  Root repo pull requests require CI to pass before merging, serving as a gate to prevent regressions. To save resources, only reviewed PRs will go through integration testing.
```

- 🔍 **Matrix Testing and Caching**  

>   Builds are optimized using matrix strategies and caching across OSes, platforms, or module combinations.

Reviewer 4: ```suggestion
  Builds are optimized using matrix strategies and caching across OSes, platforms or module combinations.
```

- 📬 **Notification and Rollback**  
  Failing tests can automatically trigger alerts, and integration changes can be rolled back based on version control and automation.

---

## 🚪 Gatekeeping and Quality Enforcement  
### *Controlled Propagation to Root*

- 🧬 **Contractual Stability**  

>   Modules declare public APIs or targets that must not be broken; test coverage is enforced to protect these contracts.

Reviewer 4: test coverage where? in the root repo or by each module separately?

- 🧷 **Change Review Requirements**  
  Any change that affects module version references in the root must undergo peer review and pass all CI checks.

- 🚫 **Preventive Merge Blocks**  
  GitHub branch protections are used to block merges when workflows fail, ensuring policy compliance.

- ✅ **Traceable Audit Logs**  
  All changes to module integration are tracked in Git history with detailed CI status, enabling forensic and compliance reviews.

---

## 🧠 Benefits & Trade-offs  
### *Balancing Agility and Structure*

- 🛤️ **Scalable for Large Teams**  
  Different teams can develop and deploy independently, without coordination delays from unrelated parts of the system.

- 🔍 **Easier Troubleshooting**  
  Failures can be isolated to individual modules or to integration logic, improving debugging speed and root cause analysis.

- 🔄 **Upfront Overhead**  

>   Requires investment in Bazel tooling (where BMW has already provided significant effort), developer training, and repository bootstrapping for new modules.

Reviewer 4: is the BMW statement really necessary here?

- 🔗 **Tighter Coordination at Integration Points**  

>   Integration still requires discipline and oversight to manage versions, rollouts, and test coverage.

Reviewer 4: ```suggestion
  Integration still requires discipline and oversight to manage versions, rollouts and test coverage.
```

---

## ✅ Conclusion & Call to Action  
### *Moving Forward with Confidence*

- 🔁 **Iterate and Test Gradually**  

>   Start by migrating a few core components into Bazel modules to gain experience before scaling out.

Reviewer 4: ```suggestion
  Start by migrating a few core components into Bazel modules to gain experience before full rollout.
```

- 📌 **Lock Down Root Governance**  

>   Enforce mandatory checks, code owners, and integration workflows in the root repository from day one.

Reviewer 4: ```suggestion
  Enforce mandatory checks, code owners and integration workflows in the root repository from day one.
```

- 🧰 **Tooling and Templates**  
  Create reusable templates and automation to help teams create compliant Bazel module repos quickly.

- 🏁 **Monitor and Optimize**  

>   Use metrics like CI test pass rates, module update frequency, and build cache efficiency to guide ongoing improvement.

Reviewer 4: ```suggestion
  Use metrics like CI test pass rates, module update frequency and build cache efficiency to guide ongoing improvement.
```
